### PR TITLE
Add Vehicle locations

### PR DESCRIPTION
**Pull request by julyto**
Description of the creation Vehicle_locations:

- adding Vehicle_location service
- creation of a structure Connector that keeps the fields as "url", "token"... for external service (in future changes the idea would be to manage the different connectors in this package)
- creation of a structure Connector that keeps the fields as "url", "token"... for Navitia (as for Connector, would be to manage the different requests with Navitia for all micro-services)

### CODE REVIEW

File: internal/vehiclelocations/navitia.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+)
+
+const (
+	URL_GET_LAST_LOAD       = "%s/status?"
+	URL_GET_VEHICLE_JOURNEY = "%s/vehicle_journeys?filter=vehicle_journey.has_code(%s)&"
+	STOP_POINT_CODE         = "gtfs_stop_code" // type code vehicle journey Navitia, the same of stop_id from Gtfs-rt
+	URL_GET_ROUTES          = "%s/lines/%s/route_schedules?direction_type=%s&from_datetime=%s"
+)
+
+type Navitia struct {
```
[the remark this answers is not in the document]
Comment (julyto) [reply internal/vehiclelocations/navitia.go:24]: Yes this structure will be used for all services. May be in new package.
It's for next step.

File: internal/utils/http.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+func GetHttpClient_(url, token, header string, connectionTimeout time.Duration) (*http.Response, error) {
```
[the remark this answers is not in the document]
Comment (julyto) [reply internal/utils/http.go:11]: Yes, I added check status http directly in GetHttpClient().
Next step all services used this method.

File: internal/vehiclelocations/Tapi_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	// Request without any parameter
+	response = VehicleLocationsResponse{}
+	c.Request = httptest.NewRequest("GET", "/vehicle_locations", nil)
+	w = httptest.NewRecorder()
+	engine.ServeHTTP(w, c.Request)
+	require.Equal(200, w.Code)
+	err = json.Unmarshal(w.Body.Bytes(), &response)
+	require.Nil(err)
+	require.NotNil(response.VehicleLocations)
+	assert.Len(response.VehicleLocations, 1)
```
[the remark this answers is not in the document]
Comment (julyto) [reply internal/vehiclelocations/Tapi_test.go:54]: Yes, I don't understand how the test went !

[the remark this answers is not in the document]
Comment (julyto) [reply internal/vehiclelocations/Tapi_test.go:54]: After verification, It is normal that there is only one element, in the Map of Vehicle_locations only one has a valid date.

File: internal/vehiclelocations/vehiclelocations.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	{
+		d.mutex.RLock()
+		defer d.mutex.RUnlock()
+
+		if d.vehicleLocations == nil {
+			e = fmt.Errorf("no vehicle_locations in the data")
+			return
+		}
+
+		// Implement filter on parameters
```
Comment (benoit-bst) [inline internal/vehiclelocations/vehiclelocations.go:79]: do you handle the case or the data is corrupt ? Like a try-catch to avoid the crash

Comment (julyto) [reply internal/vehiclelocations/vehiclelocations.go:79]: The Map must not have corrupted data, they are checked on input

**Review by kadhikari — APPROVED** [review]

File: internal/utils/http.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	client := &http.Client{Timeout: 10 * connectionTimeout}
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if token != "" && header != "" {
+		req.Header.Set("content-type", "application/x-www-form-urlencoded; param=value")
+		req.Header.Set(header, token)
+	}
```
Comment (benoit-bst) [inline internal/utils/http.go:21]:
```
else {
   return error
}
```

Comment (julyto) [reply internal/utils/http.go:21]: No, this function manages external services with or without these parameters, such as Sherbrooke for example that we query with a simple Url.

**Review by benoit-bst — APPROVED** [review]

**Pull request merged**